### ISSUE TITLE
webhook: Admission metrics are not reported by status code

### ISSUE THREAD

**Issue by ahmetb** [kubernetes-sigs/controller-runtime#2972]
_This is several bugs in one bug since it appears multiple things are broken. Apologies if it gets too complicated._

**Problem 1:** It appears that controller-runtime webhooks always respond with HTTP status code `200 OK` even if the request is completely malformed.

Example (sending a malformed request to a webhook endpoint):

```
curl -k -v https://localhost:9443/validate-ship-example-org-v1beta1-frigate
< HTTP/2 200
< ... other response headers ...
<
{"response":{"uid":"","allowed":false,"status":{"metadata":{},"message":"contentType=, expected application/json","code":400}}}
```
-----

**Problem 2:** controller-runtime webhooks do not use the `metav1.Status.Code` and always respond with HTTP status code `200 OK`. 

Example code I used to register my custom handler at `/test`:
```go
webhookServer.Register("/test", &webhook.Admission{Handler: &myHandler{}})
...

type myHandler struct{}
func (*myHandler) Handle(context.Context, admission.Request) admission.Response {
	return admission.Response{
		AdmissionResponse: admissionv1.AdmissionResponse{
			Allowed: false,
			Result: &metav1.Status{
				Code:    403,
				Message: "things suck",
				Status:  "Failure",
				Reason:  metav1.StatusReasonForbidden}}}
}
```
And here's the query I am making:
```
curl -XPOST -k -v -H "Content-Type: application/json" https://localhost:9443/test

< HTTP/2 200
...
<
{"kind":"AdmissionReview","apiVersion":"admission.k8s.io/v1","response":{"uid":"","allowed":false,"status":{"metadata":{},"status":"Failure","message":"things suck","reason":"Forbidden","code":403}}}
```

**Problem 3:** controller-runtime doesn't expose _any metrics_ about rejected requests and their `.status.code`s. This is likely because
1. we always report code=200 no matter the response (Problem 1 & 2)
2. it appears we're capping `controller_runtime_webhook_requests_total{code=...}` dimension to only `200` and `500` here: https://github.com/kubernetes-sigs/controller-runtime/blob/4cae9dfbf174fa5d49ebca007bd6f3784cf1ffb3/pkg/webhook/internal/metrics/metrics.go#L70-L76

so even if I make a faulty request, only the code=200 metric goes up:

```
controller_runtime_webhook_requests_total{code="200",webhook="/test"} 4
```

In an ideal state, I'd like to see `controller_runtime_webhook_requests_total` metric actually let me understand things like how many requests I'm denying (non-200 codes, such as 403, 400, 429, ...).

/kind bug

**Comment by troy0820** [comment]
I believe the 200 response is if the webhook can create the response.  There was an issue around this with failure policy and not sure if this helps with your issue https://github.com/kubernetes-sigs/controller-runtime/issues/2662.

**Comment by alvaroaleman** [comment]
Pretty much what troy said. I do agree that this is confusing and that it would be helpful to have metrics around the status code in the admission response.

**Comment by ahmetb** [comment]
Indeed, after creating the issue I've realized HTTP response code must be 200 regardless of the admission response (which contains the status code the apiserver would return in its response body).

Maybe we could still add a metric that parses the code out from `admission.Response`? (I understand we can't help if someone actually decides to implement their own http.Handler).

**Comment by sbueringer** [comment]
/reopen
/remove-lifecycle rotten

Adding a metric seems fine

**Comment by mattsu2020** [comment]
Hi, I'd like to work on this issue.

My understanding is that the HTTP response code should remain `200` when the webhook successfully returns an AdmissionReview response, even if the AdmissionResponse itself denies the request.

So instead of changing the HTTP status behavior, I think the improvement should be around exposing the AdmissionResponse `status.code` in metrics.

Would adding a separate metric for AdmissionResponse status codes be acceptable? For example:

```text
controller_runtime_webhook_admission_responses_total{
  webhook="...",
  allowed="false",
  admission_code="403"
}
```
